bam_samples.c

```diff
@@ -0,0 +1,459 @@
+            curr = curr->next;
+        }
+        fputc('\t', params->out);
+        if (ref == NULL) {
+            fputc('.', params->out);
+        } else {
+            fputs(curr->filename, params->out);
+        }
+    }
+    fputc('\n', params->out);
+    return 0;
+}
+
+
+/** open a sam file. Search for all samples in the @RG lines */
+static int print_samples(Params* params, const char* fname, const char* baifname) {
+    samFile *in = 0;
+    sam_hdr_t *header = NULL;
+    int n_rg;
+    int status = EXIT_SUCCESS;
+    khash_t(sm) *sample_set = NULL;
+    khint_t k;
+    int count_samples = 0;
+    int has_index = 0;
+
+    if ((sample_set = kh_init(sm)) == NULL) {
+        print_error("samples", "Failed to initialise sample hash");
+        status = EXIT_FAILURE;
+        goto end_print;
+    }
+
+    /* open sam file */
+    if ((in = sam_open_format(fname, "r", NULL)) == 0) {
+        print_error_errno("samples", "Failed to open \"%s\" for reading", fname);
+        status = EXIT_FAILURE;
+        goto end_print;
+    }
+    /* load header */
+    if ((header = sam_hdr_read(in)) == 0) {
+        print_error("samples", "Failed to read the header from \"%s\"", fname);
+        status = EXIT_FAILURE;
+        goto end_print;
+    }
+
+    /* try to load index if required */
+    if (params->test_index) {
+        hts_idx_t *bam_idx;
+        /* path to bam index was specified */
+        if (baifname != NULL) {
+            bam_idx = sam_index_load2(in, fname, baifname);
+        }
+        /* get default index */
+        else {
+            bam_idx = sam_index_load(in, fname);
+        }
+        has_index = bam_idx!=NULL;
+        if (bam_idx != NULL) hts_idx_destroy(bam_idx);
+        /* and we continue... we have tested the index file but we always test for the samples and the references */
+    }
+
+
+    /* get the RG lines */
+    n_rg = sam_hdr_count_lines(header, "RG");
+    if (n_rg > 0) {
+        int i, r, ret;
+        char* sample;
+        kstring_t sm_val = KS_INITIALIZE;
+        /* loop over the RG lines and search for the params->tag */
+        for (i = 0; i < n_rg; i++) {
+            r = sam_hdr_find_tag_pos(header, "RG", i, params->tag, &sm_val);
+            if (r < 0) continue;
+            k = kh_get(sm, sample_set, ks_str(&sm_val));
+            if (k != kh_end(sample_set)) continue;
+            sample = strdup(ks_str(&sm_val));
+            if (sample == NULL) {
+                print_error_errno("samples", "Out of memory");
+                status = EXIT_FAILURE;
+                goto end_print;
+            }
+            kh_put(sm, sample_set, sample, &ret);
+            if (ret < 0) {
+                print_error("samples", "Failed to insert key '%s' into sample_set", sample);
+                free(sample);
+                status = EXIT_FAILURE;
+                goto end_print;
+            }
+            ++count_samples;
+        }
+        ks_free(&sm_val);
+    }
+    if (count_samples == 0) {
```

Comment:
The default behaviour and output should be as generic and intuitive as possible, which IMHO is: print all the samples from a file or print nothing if there isn't any. If there is an actual need for a dot (**.**) as a replacement for a sample name, the user can inspect the manual and use the extra option to enable it.